rasa/core/policies/ted_policy.py

```diff
@@ -923,62 +909,101 @@ def _embed_label(self, label_in: Union[tf.Tensor, np.ndarray]) -> tf.Tensor:
         label = self._tf_layers[f"ffnn.{LABEL}"](label_in, self._training)
         return self._tf_layers[f"embed.{LABEL}"](label)
 
-    def _preprocess_batch(self, batch: Dict[Text, List[tf.Tensor]]) -> tf.Tensor:
-        name_features = None
-        text_features = None
-        batch_features = []
-        feats_to_combine = [f"{DIALOGUE_FEATURES}_user", f"{DIALOGUE_FEATURES}_action"]
-        for feature in feats_to_combine:
-            if not batch[feature] == []:
-                text_features = self._combine_sparse_dense_features(
-                    batch[feature], feature
-                )
-                text_features = self._tf_layers[f"ffnn.{feature}"](text_features)
-                mask = tf.cast(
-                    tf.math.equal(batch[f"{feature}_if_text"], 1), tf.float32
-                )
-                text_features = text_features * mask
-            if not batch[f"{feature}_name"] == []:
-                name_features = self._combine_sparse_dense_features(
-                    batch[f"{feature}_name"], f"{feature}_name"
-                )
-                name_features = self._tf_layers[f"ffnn.{feature}_name"](name_features)
-                mask = tf.cast(
-                    tf.math.equal(batch[f"{feature}_if_text"], -1), tf.float32
-                )
-                name_features = name_features * mask
+    def _encode_features_per_attribute(
+        self, batch: Dict[Text, Dict[Text, List[tf.Tensor]]], attribute: Text
+    ) -> Optional[tf.Tensor]:
+        if not batch[attribute]:
+            return None
 
-            if text_features is not None and name_features is not None:
-                batch_features.append(text_features + name_features)
-            else:
-                batch_features.append(
-                    text_features if text_features is not None else name_features
-                )
+        mask = batch[attribute][MASK][0]
+        attribute_features = self._combine_sparse_dense_features(
+            batch[attribute][SENTENCE], f"{attribute}_{SENTENCE}", mask=mask
+        )
+
+        if attribute in FEATURES_TO_ENCODE:
+            attribute_features = self._tf_layers[f"ffnn.{attribute}_{SENTENCE}"](
```

Comment:
let's add a `TODO` that here should be a transformer for text

Reply:
also if this ffnn contains bias, we need to multiply it by `attribute_mask` as well, otherwise it'll create nonzero values